Fix 11-year-old mis-categorized error code in OP_IF evaluation

cculianu opened the pull request:
This was introduced by commit ab9edbd6b6eb3efbca11f16fa467c3c0ef905708.

It appears the original author may have gotten tired and pasted the wrong error code into this 1 place. Every other situation where the value stack lacks the required number of arguments for the op-code, SCRIPT_ERR_INVALID_STACK_OPERATION is reported. Not so here.  

This commit fixes the situation.

EDIT: Note this turns out to be a dupe of the abandoned #30359 .

laanwj commented:
This is failing the "test each commit" check. Please squash test commits where needed to make sure that the intermediate commits pass the tests, too.

<details>
```
test/script_tests.cpp(908): Entering test case "script_json_test"
2025-03-26T09:15:59.630554Z [test] [test/util/random.cpp:48] [SeedRandomStateForTest] Setting random seed for current tests to RANDOM_CTX_SEED=be1d635f1a8e48c6048a40b25f2bd5e1d235e5ec1571241fe7157b541468c337
2025-03-26T09:15:59.630977Z [test] [init/common.cpp:151] [LogPackageVersion] Bitcoin Core version v29.99.0-da65703b810d (release build)
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: ["NOP","IF 1 ENDIF","P2SH,STRICTENC","UNBALANCED_CONDITIONAL","The following tests check the if(stack.size() < N) tests in each opcode"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: ["NOP","NOTIF 1 ENDIF","P2SH,STRICTENC","UNBALANCED_CONDITIONAL","They are here to catch copy-and-paste errors"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: ["0x03 0x635168","HASH160 0x14 0xe7309652a8e3f600f06f5d8d52d6df03d2176cc3 EQUAL","P2SH,WITNESS,MINIMALIF","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: ["0x03 0x645168","HASH160 0x14 0x0c3f8fe3d6ca266e76311ecda544c67d15fdd5b0 EQUAL","P2SH,WITNESS,MINIMALIF","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["635168",0.00000001],"","0 0x20 0xc7eaf06d5ae01a58e376e126eb1e6fab2036076922b96b2711ffbec1e590665d","P2SH,WITNESS","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["635168",0.00000001],"","0 0x20 0xc7eaf06d5ae01a58e376e126eb1e6fab2036076922b96b2711ffbec1e590665d","P2SH,WITNESS,MINIMALIF","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["645168",0.00000001],"","0 0x20 0xf913eacf2e38a5d6fc3a8311d72ae704cb83866350a984dd3e5eb76d2a8c28e8","P2SH,WITNESS","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["645168",0.00000001],"","0 0x20 0xf913eacf2e38a5d6fc3a8311d72ae704cb83866350a984dd3e5eb76d2a8c28e8","P2SH,WITNESS,MINIMALIF","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["635168",0.00000001],"0x22 0x0020c7eaf06d5ae01a58e376e126eb1e6fab2036076922b96b2711ffbec1e590665d","HASH160 0x14 0x9b27ee6d9010c21bf837b334d043be5d150e7ba7 EQUAL","P2SH,WITNESS","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["635168",0.00000001],"0x22 0x0020c7eaf06d5ae01a58e376e126eb1e6fab2036076922b96b2711ffbec1e590665d","HASH160 0x14 0x9b27ee6d9010c21bf837b334d043be5d150e7ba7 EQUAL","P2SH,WITNESS,MINIMALIF","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["645168",0.00000001],"0x22 0x0020f913eacf2e38a5d6fc3a8311d72ae704cb83866350a984dd3e5eb76d2a8c28e8","HASH160 0x14 0xdbb7d1c0a56b7a9c423300c8cca6e6e065baf1dc EQUAL","P2SH,WITNESS","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(127): error: in "script_tests/script_json_test": INVALID_STACK_OPERATION where UNBALANCED_CONDITIONAL expected: [["645168",0.00000001],"0x22 0x0020f913eacf2e38a5d6fc3a8311d72ae704cb83866350a984dd3e5eb76d2a8c28e8","HASH160 0x14 0xdbb7d1c0a56b7a9c423300c8cca6e6e065baf1dc EQUAL","P2SH,WITNESS,MINIMALIF","UNBALANCED_CONDITIONAL"]
test/script_tests.cpp(908): Leaving test case "script_json_test"; testing time: 184114us
```
</details>

cculianu committed:
Fix 11-year-old mis-categorized error code in OP_IF evaluation

This was introduced by commit ab9edbd6b6eb3efbca11f16fa467c3c0ef905708.

It appears the original author may have gotten tired and pasted the wrong
error code into this 1 place. Every other situation where the value stack
lacks the required number of arguments for the op-code,
SCRIPT_ERR_INVALID_STACK_OPERATION is reported. Not so here.

This commit fixes the situation.

Also in this commit:

- Fix script_tests to adjust to the corrected error message
- Fix p2p_invalid_tx functional test to produce the desired error message

cculianu commented:
Ok, squashed.

darosior reviewed (approved):
utACK a7b581423e44c51fb7d177c5a15fe2cc2ab8aa43

billymcbip commented:
tACK a7b581423e44c51fb7d177c5a15fe2cc2ab8aa43

Cherry-picked on latest master (2bcb3f64648ad27d4df10013b5bf0441124f1372) and ran all unit tests (including script_asset_tests).

UNBALANCED_CONDITIONAL is intended for:
- OP_ELSE/ENDIF not being preceded by OP_IF/NOTIF, or
- OP_IF/NOTIF not being succeeded by OP_ENDIF

UNBALANCED_CONDITIONAL does not make sense when OP_IF/NOTIF does not have a value to pop from the stack. INVALID_STACK_OPERATION makes sense in that case.

sedited reviewed (approved):
ACK a7b581423e44c51fb7d177c5a15fe2cc2ab8aa43

achow101 commented:
ACK a7b581423e44c51fb7d177c5a15fe2cc2ab8aa43

theuni reviewed (commented):
Post-merge utACK a7b581423e44c51fb7d177c5a15fe2cc2ab8aa43.

Very technically, this would've been an API break for libbitcoinconsensus when it existed. But now that it doesn't exist that's not really a concern :)
